pkg/client/client_v2.go

```diff
@@ -551,102 +568,118 @@ func (c *clientV2) startActions() {
 	actionResults := make(chan *proto.ActionResponse, 100)
 	go func() {
 		defer c.wg.Done()
-		for {
+		// If the initial RPC connection fails, actionRoundTrip
+		// returns immediately, so we set a timer to avoid spinlocking
+		// on the error.
+		retryTimer := time.NewTimer(1 * time.Second)
+		for c.ctx.Err() == nil {
+			c.actionRoundTrip(actionResults)
+
+			// After the RPC client closes, wait until either the timer interval
+			// expires or the context is cancelled. Note that the timer waits
+			// one second since the checkinRoundTrip was last _called_, not
+			// since it returns; immediate retries are ok after an active
+			// connection shuts down.
 			select {
+			case <-retryTimer.C:
+				retryTimer.Reset(1 * time.Second)
```

Comment:
Would it be worth using an exponential backoff here and in the checkinV2 call?

At one hand we want to reconnect to the agent relatively quickly for transient errors, on the other hand permanent failures like the deadlocks we keep seeing are just going to see the logs filled with these 1s interval error message eventually.

Reply:
I thought about it during implementation and decided it was overkill but I will poke at it again during testing...